Handle invalid WebGL MAX_TEXTURE_IMAGE_UNITS by falling back to canvas

adriandarian opened the pull request:
## Summary

This PR hardens the WebGL drawer against invalid `MAX_TEXTURE_IMAGE_UNITS` values and other context failures that can otherwise leave rendering in a broken state or crash the browser. When WebGL is not usable, OpenSeadragon now rejects that path and falls back to `CanvasDrawer` instead.

## What Changed

- validate `MAX_TEXTURE_IMAGE_UNITS` through a shared WebGL helper before relying on the context
- treat missing, zero, negative, non-finite, or context-lost values as an unusable WebGL state
- release the bad WebGL context and avoid continuing initialization when the context is invalid
- keep the existing functional WebGL support check aligned with the runtime validation path
- cover the fallback behavior in the drawer test suite, including the invalid `MAX_TEXTURE_IMAGE_UNITS` case

## Why

The original failure mode came from browsers reporting a bad WebGL state during initialization or recovery. In that case OpenSeadragon would keep going down the WebGL path even though the context was already unusable. The right behavior is the same as other WebGL failure cases in the project: do not crash, do not keep rendering with a broken context, and fall back to canvas when it is available.

## Testing

- verified the `Drawer-webgl` QUnit module passes in headless Chrome, including the invalid `MAX_TEXTURE_IMAGE_UNITS` fallback case
- build and lint continue to pass through the repo test pipeline up to the browser-runner phase in this environment

## Scope

This is focused on WebGL failure handling and regression coverage. It does not change the normal rendering path for valid WebGL contexts.

Reference: fixes the `MAX_TEXTURE_IMAGE_UNITS` crash path discussed in #2647.

adriandarian committed:
Add WebGL error handling tests and integrate into test suite

iangilman reviewed (commented):
@adriandarian this looks fantastic! Thank you for digging into this area… having better error detection, reporting, and testing for our WebGL is a big win.

I see you've added _webglFailed to keep from crashing. How does that interact with our mechanism for switching over to canvas if the WebGL doesn't work? I'm not seeing anything explicit for that, does it work implicitly?

@pearcetm since you originally wrote the WebGL drawer, I'd love to hear if you have any thoughts on this patch.

adriandarian commented:
@pearcetm @iangilman had a problem where the browser would crash on me with a webgl context lost and min texture image unit error throw in console

adriandarian committed:
feat(webgl): add mock WebGL context and update tests for error handling

adriandarian commented:
**Addressed review feedback:**

✅ **Refactored test files to reduce code duplication** (per @Copilot and @iangilman's feedback)

- Added shared WebGL mock helpers to mocks.js:
  - `createMockWebGLContext(overrides)` - Creates a complete mock WebGL context with customizable behavior
  - `createMockViewerForWebGL(props)` - Creates a mock viewer for WebGL drawer testing  
  - `mockWebGLContext()` / `restoreWebGLContext()` - Helpers for mocking/restoring `getContext`

- Refactored both test files to use the shared helpers:
  - webgl-error-handling.js: 232 → 105 lines (~55% reduction)
  - webgl-max-texture-units-simple.js: 264 → 97 lines (~63% reduction)

All 315 tests pass ✅

Other review comments were either:
- Rejected by maintainer (duplicate validation extraction - "only a couple of lines")
- Verified as correct (`_setupCanvases` method name reference)

Ready for re-review!

adriandarian committed:
fix(tests): add missing opacity property to mock viewer

adriandarian committed:
fix(tests): add missing _registerDrawer method to mock viewer

iangilman commented:
@adriandarian I haven't heard from you in a while! Are you still interested in continuing to work on these patches? 

I'm going to mark this as "up for grabs"... anyone who wants to can pick it up and take it over the finish line. That includes @adriandarian of course if you want to come back to it 😊

adriandarian committed:
fix(webgl): fall back to canvas on invalid texture units

adriandarian committed:
test(webgl): move fallback coverage into drawer suite

adriandarian committed:
Merge branch 'master' into bug/max-texture-image-unit

iangilman commented:
@adriandarian Thank you for updating all of these PRs! I imagine it makes sense for you to work on all of your PRs at the same time when you have some availability, but just to warn you, my bandwidth is such that I'm not going to be able to pay attention to all of them at the same time. I will kick off new Copilot reviews right away, and then I will take a look as I can, probably one at a time. I hope that works for you.

adriandarian committed:
fix(webgl): invalidate bad renderer setup state

adriandarian commented on src/webgldrawer.js:
```diff
@@ -220,6 +242,11 @@
             // Create unit quad once
             this._unitQuad = this.makeQuadVertexBuffer(0, 1, 0, 1);
 
+            if (!this.getMaxTextures()) {
+                $.console.warn('WebGL error: bad value for gl parameter MAX_TEXTURE_IMAGE_UNITS. Falling back to CanvasDrawer to prevent crashes.');
+                return;
+            }
```
Addressed.

`WebglContextManager.setupRenderer()` now explicitly invalidates the WebGL context when `MAX_TEXTURE_IMAGE_UNITS` is invalid instead of only logging and returning. That clears the GL handle plus the dependent renderer state (`firstPass`, `secondPass`, framebuffer, render texture, texture count, and unit quad), so later draw calls short-circuit cleanly instead of reaching a null-pass dereference path.

I also added a targeted regression test in `Drawer-webgl` that forces this exact `setupRenderer()` failure path and verifies the context and pass state are cleared.

This is now on the PR branch at `63cd942b`.

adriandarian committed:
Merge branch 'master' into bug/max-texture-image-unit

adriandarian committed:
Merge branch 'master' into bug/max-texture-image-unit

adriandarian committed:
ci: skip puppeteer download in bundle workflow
